### PR TITLE
fix: migration script can't drop constraint

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

We're trying to run the `downgrade` from `49b5a32daba5_add_report_schedules.py`, but the migration errors because the unique constraint `uq_report_schedule_name` is not found. The `upgrade` function in the migration has the creation of the constraint inside a permissive `try/except` block, so I'm assuming the upgrade partially failed for some reason and now we can't downgrade.

I modified the migration to check if the constraint exists before deleting it. The change should be harmless, because in theory the constraint should always exist. But maybe we should adopt this as a more conservative approach for `downgrade`s in migrations.

### TESTING INSTRUCTIONS

I was able to benchmark the migration:

```
Identifying models used in the migration:
- dbs (2 rows in table dbs)
- slices (114 rows in table slices)
- tables (27 rows in table tables)
Benchmarking migration
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9a30386bd74 -> 620241d1153f, update time_grain_sqla
Migration on current DB took: 0.25 seconds
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 620241d1153f -> f9a30386bd74, update time_grain_sqla
Running with at least 10 entities of each model
- Adding 8 entities to the dbs model
Processing ████████████████████████████████ 100%
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9a30386bd74 -> 620241d1153f, update time_grain_sqla
Migration for 10+ entities took: 0.39 seconds
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 620241d1153f -> f9a30386bd74, update time_grain_sqla
Running with at least 100 entities of each model
- Adding 90 entities to the dbs model
Processing ████████████████████████████████ 100%
- Adding 73 entities to the tables model
Processing ████████████████████████████████ 100%
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9a30386bd74 -> 620241d1153f, update time_grain_sqla
Migration for 100+ entities took: 0.38 seconds
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 620241d1153f -> f9a30386bd74, update time_grain_sqla
Running with at least 1000 entities of each model
- Adding 900 entities to the dbs model
Processing ████████████████████████████████ 100%
- Adding 886 entities to the slices model
Processing ████████████████████████████████ 100%
- Adding 900 entities to the tables model
Processing ████████████████████████████████ 100%
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9a30386bd74 -> 620241d1153f, update time_grain_sqla
Migration for 1000+ entities took: 0.31 seconds
Cleaning up DB

Revert DB to 620241d1153f? [y/N]: y
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Reverted

Results:

Current: 0.25 s
10+: 0.39 s
100+: 0.38 s
1000+: 0.31 s
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: scripts/benchmark_migration.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 from flask_migrate import downgrade, upgrade
 from graphlib import TopologicalSorter  # pylint: disable=wrong-import-order
 from progress.bar import ChargingBar
-from sqlalchemy import create_engine, inspect, Table
```
Comment (betodealmeida) [inline scripts/benchmark_migration.py:33]: `Table` is not used anywhere.

File: scripts/benchmark_migration.py [hunk 2]
```diff
@@ -160,7 +160,6 @@ def main(
         rows = session.query(model).count()
         print(f"- {model.__name__} ({rows} rows in table {model.__tablename__})")
         model_rows[model] = rows
-    session.close()
```
Comment (betodealmeida) [inline scripts/benchmark_migration.py:163]: Not sure why we were closing the session here, but it's not needed and actually could cause problems.

File: superset/migrations/versions/49b5a32daba5_add_report_schedules.py [hunk 3]
```diff
@@ -119,13 +120,17 @@ def upgrade():
     )
 
 
+def has_unique_constraint(constraint_name: str, table_name: str) -> bool:
+    bind = op.get_bind()
+    inspector = Inspector.from_engine(bind)
+    unique_constraints = inspector.get_unique_constraints(table_name)
+    return constraint_name in {constraint["name"] for constraint in unique_constraints}
```
Comment (betodealmeida) [inline superset/migrations/versions/49b5a32daba5_add_report_schedules.py:127]: I wonder if we should move this to `superset/utils/` and use it in all our new migrations?

**Review by nytai — APPROVED** [review]

**Review by eschutho — APPROVED** [review]
looks good.. pending a final test in sqlite.

File: superset/migrations/versions/49b5a32daba5_add_report_schedules.py (hunk 3 above)
Comment (eschutho) [reply superset/migrations/versions/49b5a32daba5_add_report_schedules.py:127]: is constraint guaranteed to have a key of "name" for every engine?

Comment (eschutho) [reply superset/migrations/versions/49b5a32daba5_add_report_schedules.py:127]: I just looked it up, that looks good.

**Comment by betodealmeida** [comment]
This is not working in SQLite, but there's also a few other migrations that are not working because they drop columns without using `op.batch`:

- superset/migrations/versions/19e978e1b9c3_add_report_format_to_report_schedule_.py
- superset/migrations/versions/2e5a0ee25ed4_refractor_alerting.py
- superset/migrations/versions/301362411006_add_execution_id_to_report_execution_.py
- superset/migrations/versions/49b5a32daba5_add_report_schedules.py
- superset/migrations/versions/585b0b1a7b18_add_exec_info_to_saved_queries.py
- superset/migrations/versions/5daced1f0e76_reports_add_working_timeout_column.py

File: superset/migrations/versions/49b5a32daba5_add_report_schedules.py (hunk 3 above)
Comment (john-bodley) [reply superset/migrations/versions/49b5a32daba5_add_report_schedules.py:127]: @betodealmeida it seems like we have similar logic like this [elsewhere](https://github.com/apache/superset/blob/62a8f2e193da3bdf1f43be085b5814be3dd476fa/superset/utils/core.py#L675-L684).

**Pull request merged**